src/main/scala/scorex/core/network/peer/PeerManager.scala

```diff
@@ -83,70 +87,68 @@ class PeerManager(settings: ScorexSettings, timeProvider: NetworkTimeProvider) e
   private def isSelf(address: InetSocketAddress, declaredAddress: Option[InetSocketAddress]): Boolean = {
     // TODO: should the peer really be considered the same as self iff one of the following conditions hold?? Check carefully.
     settings.network.bindAddress == address ||
-    settings.network.declaredAddress.exists(da => declaredAddress.contains(da)) ||
-    declaredAddress.contains(settings.network.bindAddress) ||
-    settings.network.declaredAddress.contains(address)
+      settings.network.declaredAddress.exists(da => declaredAddress.contains(da)) ||
+      declaredAddress.contains(settings.network.bindAddress) ||
+      settings.network.declaredAddress.contains(address)
   }
 
+  private var lastIdUsed = 0
 
   private def peerCycle: Receive = {
-    case Connected(newPeer@ConnectedPeer(remote, _)) =>
-      if (peerDatabase.isBlacklisted(newPeer.socketAddress)) {
+    case Connecting(remote, direction) =>
+      val peerHandlerRef = sender()
+
+      if (peerDatabase.isBlacklisted(remote)) {
         log.info(s"Got incoming connection from blacklisted $remote")
       } else {
-        connectedPeers += newPeer -> None
-        if (connectingPeer.contains(remote)) {
-          log.info(s"Connected to $remote. ${connectedPeers.size} connections are open")
-          connectingPeer = None
+        val refuse =
+          if (direction == Incoming) false
+          else if (connectingPeers.contains(remote)) {
+            log.info(s"Connecting to $remote")
+            false
+          } else {
+            log.info(s"Already connected peer $remote trying to connect, going to disconnect it")
```

Comment:
Are we really going to disconnect it? Or just refuse the second connection? I think this should be clarified.

Reply:
done, new message is "Already connected peer $remote trying to connect, going to drop the duplicate connection" (so original connection is still alive)